api/server/router/local/image.go

```diff
@@ -584,25 +635,49 @@ func (s *router) getImagesSearch(ctx context.Context, w http.ResponseWriter, r *
 		return err
 	}
 	var (
-		config      *cliconfig.AuthConfig
+		configs     = make(map[string]cliconfig.AuthConfig)
 		authEncoded = r.Header.Get("X-Registry-Auth")
 		headers     = map[string][]string{}
+		term        = r.Form.Get("term")
 	)
 
 	if authEncoded != "" {
 		authJSON := base64.NewDecoder(base64.URLEncoding, strings.NewReader(authEncoded))
-		if err := json.NewDecoder(authJSON).Decode(&config); err != nil {
+		if err := json.NewDecoder(authJSON).Decode(&configs); err != nil {
 			// for a search it is not an error if no auth was given
 			// to increase compatibility with the existing api it is defaulting to be empty
-			config = &cliconfig.AuthConfig{}
+			configs = make(map[string]cliconfig.AuthConfig)
 		}
 	}
+	// maybe client just sends one auth config
+	// try to resolve just one auth config...
+	authConfig := cliconfig.AuthConfig{}
+	if len(configs) == 0 {
+		if authEncoded != "" {
+			authJSONSingle := base64.NewDecoder(base64.URLEncoding, strings.NewReader(authEncoded))
+			if err := json.NewDecoder(authJSONSingle).Decode(&authConfig); err != nil {
+				// for a pull it is not an error if no auth was given
+				// to increase compatibility with the existing api it is defaulting to be empty
+				authConfig = cliconfig.AuthConfig{}
+			}
+		}
+	}
+
+	// this is the case when we fully qualify images
+	repoInfo, err := registry.ParseRepositoryInfo(term)
+	if err != nil {
+		return err
+	}
+	if len(configs) == 0 {
+		configs[repoInfo.Index.Name] = authConfig
+	}
+
```

Comment:
There's too much rendundancy in this file. Could you put the authparsing into a helper?